File: samples/FileUploaderExample/Vci.FileUploader/UploaderControl.cs
/*
Copyright 2003-2009 Virtual Chemistry, Inc. (VCI)
http://www.virtualchemistry.com
Using .Net, Silverlight, SharePoint and more to solve your tough problems in web-based data management.

Author: Peter Coley
*/

using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Linq;
using System.IO;
using System.Text;
using System.Web;
using System.Web.UI;
using System.Web.UI.WebControls;
using System.Web.UI.HtmlControls;

using Vci.Core;

namespace Vci.FileUploader
{
    /// <summary>
    /// Convenient wrapper on the silverlight uploader control that facilitates passing initialization parameters,
    /// and initializes javascript object wrapper that can be used to interact with the control.
    /// </summary>
    public class UploaderControl : WebControl
    {
        private bool _multiSelect = true;

        private string _scriptPath = "~/scripts/uploadercontrol.js";
        private string _xapFile = "~/ClientBin/SilverlightFileUploader.xap";
        private string _handlerPath = "~/UploaderControlHandler.ashx";

        /// <summary>
        /// Maximum number of concurrent uploads.  Defaults to 2.
        /// </summary>
        public int MaxUploads { get; set; }

        /// <summary>
        /// Maximum size of an uploaded file in KB.  Default is 102400 KB (100 MB).
        /// </summary>
        public int MaxFileSizeKB { get; set; }

        /// <summary>
        /// Chunk size in MB.  Defaults to 3 MB.
        /// </summary>
        public int ChunkSizeMB { get; set; }

        /// <summary>
        /// Filter for the types of files that can be uploaded.
        /// </summary>
        public string FileFilter { get; set; }

        /// <summary>
        /// If true, multiple files can be uploaded and chosen at once.  Defaults to true.
        /// </summary>
        public bool MultiSelect { get { return _multiSelect; } set { _multiSelect = value; } }

        /// <summary>
        /// If specified, this is a class that implements IUploadedFileProcessor.  This Type's ProcessFile
        /// method will be invoked on the server for each file that is successfully uploaded.
        /// </summary>
        public string UploadedFileProcessorType { get; set; }

        /// <summary>
        /// Location of the uploadercontrol.js file; defaults to ~/scripts/uploadercontrol.js.
        /// </summary>
        public string ScriptPath { get { return _scriptPath; } set { _scriptPath = value; } }

        /// <summary>
        /// Path to the xap file, defaults to ~/ClientBin/SilverlightFileUploader.xap.
        /// </summary>
        public string XapPath { get { return _xapFile; } set { _xapFile = value; } }

        /// <summary>
        /// Path to the http handler that will receive uploaded file data, defaults to ~/UploaderControlHandler.ashx.
        /// </summary>
        public string HandlerPath { get { return _handlerPath; } set { _handlerPath = value; } }

        /// <summary>
        /// The user context is passed as ContextParam to the upload file processor.  You may also set this value in
        /// javascript via the setUserContext property before uploading has begun.
        /// </summary>
        public string UserContext { get; set; }

        protected override void OnLoad(EventArgs e)
        {
            base.OnLoad(e);

            ClientScriptManager csm = Page.ClientScript;

            if (!csm.IsClientScriptIncludeRegistered("uploadercontrol"))
                csm.RegisterClientScriptInclude("uploadercontrol", ResolveClientUrl(_scriptPath));
        }

        protected override void Render(HtmlTextWriter writer)
        {
            ClientScriptManager csm = Page.ClientScript;

            string source = ParseUrlProperty(_xapFile);

            Dictionary<string, string> initParamList = new Dictionary<string, string>();
            initParamList["UploadHandlerPath"] = ParseUrlProperty(_handlerPath);

            if (MaxUploads > 0)
                initParamList["MaxUploads"] = MaxUploads.ToString();

            if (MaxFileSizeKB > 0)
                initParamList["MaxFileSizeKB"] = MaxFileSizeKB.ToString();

            if (ChunkSizeMB > 0)
                initParamList["ChunkSizeMB"] = ChunkSizeMB.ToString();

            if (!string.IsNullOrEmpty(FileFilter))
                initParamList["FileFilter"] = FileFilter;

            if (!string.IsNullOrEmpty(UploadedFileProcessorType))
                initParamList["UploadedFileProcessorType"] = UploadedFileProcessorType;

            initParamList["MultiSelect"] = _multiSelect.ToString();

            string sInitParams = SilverlightUtility.BuildInitParamsString(initParamList);

            // NOTE: this uses "windowless" mode by default; feel free to change this as you see fit... in an app like
            // this that is not particularly intensive on the UI side of things, using windowless mode makes embedding
            // the control into your pages much more manageable, particularly in firefox -- avoids a lot of the issues
            // with setting dimensions on containers and using display:none

            writer.Write(@"
<object id=""silverlight" + this.ID + @""" data=""data:application/x-silverlight-2,"" type=""application/x-silverlight-2"" width=""100%"" height=""100%"">
    <param name=""source"" value=""" + source + @"""/>
    <param name=""minRuntimeVersion"" value=""2.0.31005.0"" />
    <param name=""autoUpgrade"" value=""true"" />
    <param name=""initParams"" value=""" + sInitParams + @""" />
    <param name=""windowless"" value=""true"" />
    <param name=""onLoad"" value=""UploaderControl_onLoad_" + this.ID + @""" />
    <a href=""http://go.microsoft.com/fwlink/?LinkID=124807"" style=""text-decoration: none;"">
        <img src=""http://go.microsoft.com/fwlink/?LinkId=108181"" alt=""Get Microsoft Silverlight"" style=""border-style: none""/>
    </a>
</object>
<iframe style='visibility:hidden;height:0;width:0;border:0px'></iframe>
");

            string script = @"
function UploaderControl_onLoad_" + this.ID + @"() {
    UploaderControl.getInstance(""" + this.ID + @""").onLoad();
}

function UploaderControl_uploadStarted_" + this.ID + @"(sender, e) {
    UploaderControl.getInstance(""" + this.ID + @""").uploadStarted(sender, e);    
}

function UploaderControl_singleFileUploadFinished_" + this.ID + @"(sender, e) {
    UploaderControl.getInstance(""" + this.ID + @""").singleFileUploadFinished(sender, e);    
}

function UploaderControl_allFilesFinished_" + this.ID + @"(sender, e) {
    UploaderControl.getInstance(""" + this.ID + @""").allFilesFinished(sender, e);    
}

function UploaderControl_errorOccurred_" + this.ID + @"(sender, e) {
    UploaderControl.getInstance(""" + this.ID + @""").errorOccurred(sender, e);    
}

UploaderControl.getInstance(""" + this.ID + @""").initialize(" + (UserContext == null ? "null" : "\"" + UserContext + "\"") + @");
";
            csm.RegisterStartupScript(this.GetType(), this.ID, script, true);
        }

        /// <summary>
        /// Replaces ~ with the application path
        /// </summary>
        /// <param name="Url"></param>
        /// <returns></returns>
        private string ParseUrlProperty(string Url)
        {
            string ret = Url;
            if (Url.StartsWith("~"))
            {
                string appPath = Page.Request.ApplicationPath;
                if (appPath == "/") appPath = "";
                ret = appPath + Url.Substring(1);
            }
            return ret;
        }
    }
}

File: samples/FileUploaderExample/SilverlightFileUploader.Web/scripts/uploadercontrol.js
/*
Copyright 2003-2009 Virtual Chemistry, Inc. (VCI)
http://www.virtualchemistry.com
Using .Net, Silverlight, SharePoint and more to solve your tough problems in web-based data management.

Author: Peter Coley
*/

if(typeof(UploaderControl) == "undefined" || UploaderControl == null) {

UploaderControls = {};

UploaderControl = function(id) {
	
	if(arguments.length > 0)
		this.init(id);
};

UploaderControl.getInstance = function(id) {
	var x = UploaderControls[id];
	if(x == null)
		x = UploaderControls[id] = new UploaderControl(id);
	return x;
};

UploaderControl.eventObjects = {};

UploaderControl.prototype = {
	
	id : null,
	silverlightClientId : null,
	loaded : false,
	uploadInProgress : false,
	userContext : null,
	
	init : function(id) {
	
	    this.id = id;
	    this.events = {};
	    
		UploaderControl.eventObjects[this.id] = this;
		
		//
		// These are the available js events that you can handle on the UploaderControl.  See attachEvent for details on how to use.
		//
		this.events["onuploadstarted"] = null; // the user has started uploading some files; this is your last chance to set the user context
		this.events["onfileuploaded"] = null; // (fileGuid, fileName, fileSize) called for each file when it has completed uploading
		this.events["onallfilesuploaded"] = null; // all files have completed uploading (or been canceled or removed from the list)
		this.events["onerror"] = null; // (errorMessage)
	},

	initialize : function(userContext) {

        this.userContext = userContext;
        this.silverlightClientId = "silverlight" + this.id;

        // it's possible that the control can finish loading before the load handling js function is written to the page, so do a check
        if(!this.loaded) {
            try {
                var obj = document.getElementById(this.silverlightClientId);
                if(obj && obj.Content && obj.Content.Files)
                    this.onLoad();
            } catch(ex) {}
        }
    },
    
    //
    // Public methods.
    //
    
    setUserContext : function(ctx) {
    /// <summary>
    /// A string that is sent as the Context parameter to server-side file upload processors.  Must be set before
    /// file uploading has begun.
    /// </summary>
    
        this.userContext = ctx;
    },

	attachEvent : function(eventName, func, context) {
	/// <summary>
	/// Public method for attaching event handlers to this control.
	/// </summary>
	/// <param name="eventName">the name of the event</param>
	/// <param name="func">function pointer to the event handler</param>
	/// <param name="context">context to use when calling func (becomes the 'this' keyword within your function)</param>
	/// <example>
	/// var uploader = UploaderControl.getInstance("upload1");
	/// uploader.attachEvent("onfileuploaded", myFunc);
	/// ...
	/// function myFunc(fileGuid, fileName, fileSize) {
	///     ...
	/// }
	/// </example>

		if(context == null) context = this;
		if(func == null)
			this.events[eventName] = null;
		else
			this.events[eventName] = {func : func, context : context};
	},
    
    
    //
    // below are some example utility methods that access methods on the silverlight object
    //
    
    clearFileList : function() {
    
        if(this.loaded) {
            var obj = document.getElementById(this.silverlightClientId);
            try {
                if(obj && obj.Content && obj.Content.UploaderControl)
                    obj.Content.UploaderControl.ClearFileList();
            } catch(ex) {}
        }
    },
    
    getFileCount : function() {

        var ret = -1;
        if(this.loaded) {
            var obj = document.getElementById(this.silverlightClientId);
            try {
                if(obj && obj.Content && obj.Content.Files)
                    ret = obj.Content.Files.TotalFilesSelected;
            } catch(ex) {}
        }
        
        return ret;
    },
    
    
    //
    // private methods for hooking stuff up to the silverlight control
    //
    
    onLoad : function() {
    
        this.loaded = true;
        
        var obj = document.getElementById(this.silverlightClientId);
        obj.Content.UploaderControl.UploadStarted = window["UploaderControl_uploadStarted_" + this.id];
        obj.Content.Files.SingleFileUploadFinished = window["UploaderControl_singleFileUploadFinished_" + this.id];
        obj.Content.Files.AllFilesFinished = window["UploaderControl_allFilesFinished_" + this.id];
        obj.Content.Files.ErrorOccurred = window["UploaderControl_errorOccurred_" + this.id];
    },
    
    uploadStarted : function(sender, e) {
    
        this.invokeEventHandler("onuploadstarted");
        
        this.uploadInProgress = true;
        var obj = document.getElementById(this.silverlightClientId);
        obj.Content.Files.UserContextParameter = this.userContext;        
    },
    
    singleFileUploadFinished : function(sender, e) {
    
        this.invokeEventHandler("onfileuploaded", e.FileGuid, e.FileName, e.FileSize);
    },
    
    allFilesFinished : function(sender, e) {
    
        this.uploadInProgress = false;
        this.invokeEventHandler("onallfilesuploaded");
    },
    
    errorOccurred : function(sender, e) {
    
        this.invokeEventHandler("onerror", e.ErrorMessage);
    },
    
	getEventHandler : function(eventName) {
		return this.events[eventName];
	},
	
	invokeEventHandler : function(eventName) {

		if((handler = this.events[eventName]) != null) {
			var args = new Array();
			for(var i = 1; i < arguments.length; i++)
				args.push(arguments[i]);
			return handler.func.apply(handler.context, args);
		}
		
		return null;
	}    
};

}
